sys: util: Prevent multiple definitions of ARRAY_SIZE

mfischer opened the pull request:
Wrap ARRAY_SIZE macro with an #ifndef ARRAY_SIZE to prevent multiple definition issues when dealing with third party libraries that have their own definitions.

Fixes #51371

Signed-off-by: Moritz Fischer <moritzf@google.com>

mfischer committed:
sys: util: Prevent multiple definitions of ARRAY_SIZE

Prefix ARRAY_SIZE macro with an #undef ARRAY_SIZE to prevent
multiple definition issues when dealing with third party
libraries that have their own definitions.

Signed-off-by: Moritz Fischer <moritzf@google.com>

dcpleung commented:
Similar PR was closed without merging: https://github.com/zephyrproject-rtos/zephyr/pull/49831

mfischer commented:
> Similar PR was closed without merging: #49831

What is the reasoning to not allow this, but do the same thing for other macros? Every single implementation of `ARRAY_SIZE` I've seen in the wild looked exactly the same, and expecting every library out there to carry patches for Zephyr is pretty unreasonable in my book.

dcpleung commented:
Please read the comments of the linked PR and see if you can satisfy what has been voiced there.

cfriedt reviewed (changes requested):
Add tests here, please.

https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/unit/util/test.inc

Unless it can be proven that there is some sort of security threat by doing so, I see absolutely no reason not to ack here.

On a side note, we probably do want to have a non-unit test that explicitly causes a crash to validate failure modes. IIRC, it is not possible to do that with a zephyr "unit test" (because they run as native_posix).

mfischer commented:
> main/tests/unit/util/test.inc

There are tests in there that use `ARRAY_SIZE`. I can add a 
```
void run_ARRAY_SIZE(void)
{
  size_t array[] = {1,2,3,4};
  zassert_equal(4, ARRAY_SIZE(array));
}

```

If you want something explicit.

mfischer committed:
tests: unit: Add explicit ARRAY_SIZE tests

This adds a test that explicitly (rather than implicitly)
tests the ARRAY_SIZE(x) macro.

Signed-off-by: Moritz Fischer <moritzf@google.com>

de-nordic reviewed (changes requested):
This has been discussed here https://github.com/zephyrproject-rtos/zephyr/pull/49831, as @dcpleung has noted, and abandoned by @nordicjm, as a result of the discussion.

Making this conditional may make the macro behave differently across compilation units, when different definitions are provided.
Do we really want to allow that?

de-nordic commented:
> 
> This is for 3rd-party code compatibility, so simply avoid a compiler warning that will be promoted to error in twister.
> 

Avoiding warning does not fix problem it makes it just creep in the shadows.
Avoiding  warnings can be done by running sed/awk in pipeline and filtering them out: works the same, does not fix the problem.
I have already spent some time in my life on debugging things that could have been found in warnings, warnings suppressed because they would cause warning count go above level accepted by quality - no difference from here.

I have also gone through something like
```
#undef SUCCESS
#define SUCCESS 0
```
which is basically "happy debugging suckers" case from bash.org, that makes me reluctant to allow different definitions depending on the order of header inclusions.

> We all know the correct fix for this - i.e. use a namespaced `Z_ARRAY_SIZE()`. However, there is constant pushback against that.
> 

Then, the warning is the argument in the discussion. 

> As an intermediate step to replacing `ARRAY_SIZE()` everywhere, deprecate it, create a `CONFIG_DEFINE_ARRAY_SIZE` that defaults to `y`, such that
> 
> ```
> #ifdef CONFIG_DEFINE_ARRAY_SIZE
> #define ARRAY_SIZE(x) Z_ARRAY_SIZE(x)
> #endif
> ```
> 
> Eventually, removing the CONFIG / define.

That would be great.

gmarull commented:
> We all know the correct fix for this - i.e. use a namespaced `Z_ARRAY_SIZE()`. However, there is constant pushback against that.
> 

I wish Zephyr had a namespace. IMO, it's a mistake not to do that when using lots of 3rd party code in-tree (HALs et al.).

de-nordic commented:
> 
> I wish Zephyr had a namespace. IMO

Me too.

> it's a mistake not to do that when using lots of 3rd party code in-tree (HALs et al.).

Yes. Having our own namespace kinda lets us move out of the way, in most of cases as `ARRAY_SIZE`(TM)(R)(C) is very  generic name.

I must admit that the 3rd party code is not without blame, because using generic names for things is just asking for collisions.
It is obvious that user needs to include headers of some lib to use its API, but if these bring redefinitions of things like `ARRAY_SIZE`, `OK`, `FALSE`, `SUCCESS` or whatever you thing about now, interface of the library probably exposes something that should be internal, is missing prefixes, or contains inclusions of headers it should not take.

cfriedt commented:
@mfischer - well, how would you feel ripping the bandaid off?

mfischer commented:
> @mfischer - well, how would you feel ripping the bandaid off?

Sure, this week is busy. I'll see when I get around to it. During -rc nothing will be merged anyways :)

SebastianBoe commented:
IMHO it is acceptable that you get a corrupt build when you include a third-party definition of ARRAY_SIZE that does not return the array size.

There is already prior-art for ifndef with MAX and MIN, so it should be permitted to also do this for ARRAY_SIZE.

Introducing a namespace for MAX, MIN, ARRAY_SIZE and everything else in util.h should be out-of-scope of this PR.

I agree we should not allow ifndef of SUCCESS, because that is obviously ambiguous.

But MAX, MIN, and ARRAY_SIZE are well defined.

cfriedt reviewed (dismissed):
I'm also fine with this as it is. Agree it would be better to namespace macros / add configs and deprecations in a separate PR.

mfischer commented:
> This has been discussed here #49831, as @dcpleung has noted, and abandoned by @nordicjm, as a result of the discussion.
> 
> Making this conditional may make the macro behave differently across compilation units, when different definitions are provided. Do we really want to allow that?

It seems like the consensus is to do this in two steps. As a first step let's get this merged.

I'll deal with the Z_ARRAY_SIZE, Z_MIN, ... in follow up PRs

cfriedt commented:
@de-nordic - do you still need changes in this PR? If not, please unblock.

de-nordic commented:
> IMHO it is acceptable that you get a corrupt build when you include a third-party definition of ARRAY_SIZE that does not return the array size.

Source code that redefines system/compiler provided macros should do that explicitly by `undef`, and have it documented. I do not think it is good, for the system, to walk out of the way.

`MIN/MAX` are non-standard macros, not defined by C standard, and may or may not be provided by systems.

> 
> There is already prior-art for ifndef with MAX and MIN, so it should be permitted to also do this for ARRAY_SIZE.
> 

`ifndef` thing is not eniterly true:
For example FreeBSD:
 https://github.com/freebsd/freebsd-src/blob/ffbc2a58b13a9239269e661fe23f86c587cc7638/sys/sys/param.h#L329-L331
just defines `MIN`/`MAX`, no checks.

Glib does not even bother to `#ifndef`, it just goes for `#undef`
https://github.com/GNOME/glib/blob/2843eef4a4fdccbef5782b8a5b7cd121c8b91ca0/glib/gmacros.h#L887-L894

Linux kernel sources do not have "global" `MAX` definition.

GNU libc, https://codebrowser.dev/glibc/glibc/misc/sys/param.h.html#101, doesn't bother to `#ifndef` either.

> 
> I agree we should not allow ifndef of SUCCESS, because that is obviously ambiguous.
> 
> But MAX, MIN, and ARRAY_SIZE are well defined.

Not by system. Looking at Linux kernel sources, there is no `MAX`, `MIN`, `ARRAY_SIZE` definition provided by kernel headers, system interface headers.
Looking at FreeBSD, the `ARRAY_SIZE` is only defined by subsystems, for local use. Also there is also something named `nitems` where some `ARRAY_SIZE` definitions use like
```
#define ARRAY_SIZE(x) nitems(x)
```

which is generally defined as:
```
#define nitems(x)       (sizeof((x)) / sizeof((x)[0]))
```
basically what ARRAY_SIZE(x) does, right? Except for this one definition: https://github.com/freebsd/freebsd-src/blob/main/tools/build/cross-build/include/common/sys/param.h#L70-L74.
Seems that results will differ, depending on header order.

If you include to your code some library that redefines stuff provided by system headers, or toolchain headers, it is on you to `undef` or change header order; that lib needs fixing.

Including header files should not be art of sudoku, where you are only sure that Zephyr headers could be pushed down below everything else, because they are sorry to have been included and will move out of others way just to avoid arguing. 

I am more willing into doing this Glib way: "I am system header better put me on top because I `undef` your stuff without asking".
And then everything else has do do `undef` explicitly, or `ifndef ` the `sys/util.h` provided macros.

This is system header, it is used in Zephyr code base and whatever identifiers it provides, when included into subsytem code, driver code, and so on, it should provide identifier with definition that is within `sys/util.h'; this should not be dependent on header order.

> Introducing a namespace for MAX, MIN, ARRAY_SIZE and everything else in util.h should be out-of-scope of this PR.

So what is scope of this PR, because now it seems that it will become permanent workaround for not being able to agree on namespace, although I am not even sure if we need one, as I have mentioned above: this is system header; it it collides in internal subsystem you are writing - the decision is on you to not include it or `undef`,  if it is in your app, that is completely on you, because this is internal header.

de-nordic reviewed (commented):
I will not be blocking this although I would go rather with `undef`, same way the glib does, than "I am sorry I got into your way" `ifndef`.
More on that here https://github.com/zephyrproject-rtos/zephyr/pull/50239#issuecomment-1265195103.
Not blocking, not approving.

SebastianBoe commented:
Changing MAX and MIN in sys/util.h from using ifndef to undef as glib does sound reasonable and in-scope for this PR.

Thank you for the thorough review.

mfischer commented:
> I will not be blocking this although I would go rather with `undef`, same way the glib does, than "I am sorry I got into your way" `ifndef`. More on that here [#50239 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/50239#issuecomment-1265195103). Not blocking, not approving.

So do we want to change *all* of them to
```
#undef ARRAY_SIZE
#define ARRAY_SIZE ...

#undef MIN
#define MIN...
```
instead?

de-nordic commented:
@mfischer Also add release notes commit with comment in macro behaviour. I think that people should be notified that now the `sys/util.h` headers stands its ground, regarding the definitions.

mfischer committed:
sys: util: Take stronger stance on MIN()/MAX()

Instead of guarding with #ifndef against previous definitions
use a stronger stance and #undef previous definitions.

Signed-off-by: Moritz Fischer <moritzf@google.com>

mfischer committed:
doc/release-note-3.3: Add notes about MAX/MIN/ARRAY_SIZE

Added summary of new behavior for MAX(), MIN() and ARRAY_SIZE().

Signed-off-by: Moritz Fischer <moritzf@google.com>

mfischer commented:
> @mfischer Also add release notes commit with comment in macro behaviour. I think that people should be notified that now the `sys/util.h` headers stands its ground, regarding the definitions.

Done. PTAL

andyross reviewed (changes requested):
Is this back from the dead?  Can we not do this?  This kind of trickery produces output code that depends on header inclusion order (whoever ends up in the parse first wins).  That's just a bad idea, and in any case won't work without doing the same trickery in the colliding library (i.e. after merging this, you will STILL get a collision if you include the Zephyr header first).

Let's fix symbol collisions like grownups and not use the same symbol names.

It would be best to fix this in the other code, frankly.  Zephyr is the OS and should take precedence over random libraries if we really have to have the fight (c.f. ARRAY_SIZE is also a symbol in linux).

But if we have to do it here, pick one of the Zephyr prefixes (SYS_ would probably be best) and use that.

andyross commented:
Oh, heh, not back from the dead, just two folks fixing the same thing.  Is it a collision with the same library?

gmarull commented:
related: https://github.com/zephyrproject-rtos/hal_nxp/pull/196

de-nordic commented:
>  (i.e. after merging this, you will STILL get a collision if you include the Zephyr header first).

@andyross What do you mean by collision? Current version of this PR undefs any preceding definitions of macros and then defines them. Zephyr wins.

> 
> Let's fix symbol collisions like grownups and not use the same symbol names.
> 

The problem is that things that should not be in interface headers, like definitions with generic names, are there and they cause conflicts.

> It would be best to fix this in the other code, frankly. Zephyr is the OS and should take precedence over random libraries if we really have to have the fight (c.f. ARRAY_SIZE is also a symbol in linux).

Agree; in my comment https://github.com/zephyrproject-rtos/zephyr/pull/50239#issuecomment-1265195103 I have unfortunately missed definition in include/linux/kernel.h, sorry for that.

> But if we have to do it here, pick one of the Zephyr prefixes (SYS_ would probably be best) and use that.

I am no longer sure we need that. sys/util.h is, IMHO, internal util header; other systems to not bother add prefixes to their definitions.

stephanosio commented:
> @andyross What do you mean by collision? Current version of this PR undefs any preceding definitions of macros and then defines them. Zephyr wins.

What if you include another (third-party) header that does the same thing after the Zephyr header?

These re-definitions only fail because the re-definition is different from the original definition (compiler will not complain for the identical re-definition) -- that IS a problem.

> It would be best to fix this in the other code

Agreed.

> > But if we have to do it here, pick one of the Zephyr prefixes (SYS_ would probably be best) and use that.
> 
> I am no longer sure we need that. sys/util.h is, IMHO, internal util header; other systems to not bother add prefixes to their definitions.

I think `SYS_` is still useful. By prefixing all these macros and "claiming" the `SYS_` namespace (reasonable since Zephyr is an operating **system**), we can then clearly say any third-party code defining anything that starts with `SYS_` are wrong.

cfriedt commented:
> But if we have to do it here, pick one of the Zephyr prefixes (SYS_ would probably be best) and use that.

I kind of like `Z_` for a namespace prefix and we seem to be using that all over the place already.

de-nordic commented on include/zephyr/sys/util.h:
```diff
@@ -105,6 +106,7 @@ extern "C" {
  */
+#undef ARRAY_SIZE
 #define ARRAY_SIZE(array) \
 	((size_t) (IS_ARRAY(array) + (sizeof(array) / sizeof((array)[0]))))
```
> In practice this means ARRAY_SIZE implementation will depend on the including order: BAD. Note that since Zephyr seems to be worried about safety, this violates MISRA 20.5 (advisory) as well.

Valid arguments.

> 
> I think we should either:
> 
>     * Fix offending libraries/3rd party code

Yes.

> 
>     * Use a namespace in Zephyr, to protect us from 3rd party code.

There is always chance for collision, and system should not move out of the way.
Choosing SYS_ or Z_ prefix will fix a problem until other system gets the same idea and some interface header for some library, hal or other thing  does the #ifndef/#define thing in interface header again.

Only way to fix header order problem is to avoid having definitions, in headers, that should not be there.

sys/util.h is included specifically for the definitions it provides, so it is not guilty of redefinitions.

Question: how many Zephyr headers include sys/util.h for no reason?

cfriedt commented:
I will also +1 this change if it becomes a tree-wide change to `Z_ARRAY_SIZE()` as long as we deprecate `ARRAY_SIZE()` and activate it with a y-defaulted Kconfig. That way, if a particular build sees a conflict in the definition of `ARRAY_SIZE()`, it can just be deactivated with a Kconfig. Also does not break modules, since it is y-defaulted.

This change is practically as simple as running a grep, awk, and sed one-liner.

dleach02 commented:
I'm stepping into this thread since the NXP HAL is one of the modules/libraries that @gmarull is trying to change. While the change proposed is not that big, the problem I have is that I have to propagate this change back to the source or I'm burdened with the technical debt of maintaining a patch that has to be applied. 

Previously, it was header include order that prevented this problem from happening and partially because the NXP HAL definition wrapped the ARRAY_SIZE with an #ifndef. We don't seem to want to do that with the Zephyr version to, I assume, ensure that the system build gets the functionality being defined by our version. 

The other solution being discussed is to fix this via a namespace prefix. 

My vote is for a namespace fix. ARRAY_SIZE is a super common define in the embedded space. Below is a quick scan of our current modules/libraries on the frequency of its definition so rather then forcing changes on every HAL/Module/Lib we pull into Zephyr we should just ensure a separate name space for Zephyr's implementation of these various common defines (or just use the #ifndef wrapper).

\bootloader\mcuboot\boot\boot_serial\src\zcbor_common.h:https://github.com/zephyrproject-rtos/zephyr/pull/333
\modules\audio\sof\tools\fuzzer\fuzzer.h:https://github.com/zephyrproject-rtos/zephyr/pull/33
\modules\hal\espressif\components\bt\esp_ble_mesh\mesh_common\include\mesh_util.h:https://github.com/zephyrproject-rtos/zephyr/pull/59
\modules\hal\espressif\components\wpa_supplicant\src\utils\common.h:https://github.com/zephyrproject-rtos/zephyr/pull/443
\modules\hal\telink\tlsr9\ble\common\utility.h:https://github.com/zephyrproject-rtos/zephyr/pull/104
\modules\lib\chre\chpp\include\chpp\macros.h:https://github.com/zephyrproject-rtos/zephyr/pull/28
\modules\lib\loramac-node\src\boards\mcu\sam121\hal\utils\include\utils.h:https://github.com/zephyrproject-rtos/zephyr/pull/65
\modules\lib\zcbor\include\zcbor_common.h:https://github.com/zephyrproject-rtos/zephyr/pull/74
\modules\tee\tf-a\trusted-firmware-a\include\lib\utils_def.h:https://github.com/zephyrproject-rtos/zephyr/pull/14
\modules\tee\tf-m\trusted-firmware-m\secure_fw\include\array.h:https://github.com/zephyrproject-rtos/zephyr/pull/15 (a number of different files in TF-M define this)

de-nordic commented:
The zcbor has already been updated with replacement of `ARRAY_SIZE` by `ZCBOR_ARRAY_SIZE`
https://github.com/zephyrproject-rtos/zcbor/blob/main/include/zcbor_common.h#L85-L87
The mcuboot has not yet been updated with the new code.

carlescufi reviewed (changes requested):
See conversation here: https://discord.com/channels/720317445772017664/1014241011989487716/1032623375228608522

carlescufi commented:
The approach in this PR has been rejected in the different forums where this has been discussed, and instead a completely different strategy is under review:
https://github.com/zephyrproject-rtos/zephyr/pull/51963

yperess reviewed (approved):
LGTM on the tests/ end
